diff --git a/util/libcrypto.num b/util/libcrypto.num
index 30ae5986bf0f..204236dc85bc 100644
--- a/util/libcrypto.num
+++ b/util/libcrypto.num
@@ -5425,8 +5425,8 @@ ASN1_item_d2i_ex                        5552	3_0_0	EXIST::FUNCTION:
 ASN1_TIME_print_ex                      5553	3_0_0	EXIST::FUNCTION:
 EVP_PKEY_get0_provider                  5554	3_0_0	EXIST::FUNCTION:
 EVP_PKEY_CTX_get0_provider              5555	3_0_0	EXIST::FUNCTION:
-OPENSSL_strcasecmp                      5556	3_0_3	EXIST::FUNCTION:
-OPENSSL_strncasecmp                     5557	3_0_3	EXIST::FUNCTION:
+OPENSSL_strcasecmp                      5556	3_0_1	EXIST::FUNCTION:
+OPENSSL_strncasecmp                     5557	3_0_1	EXIST::FUNCTION:
 OSSL_CMP_CTX_reset_geninfo_ITAVs        5558	3_0_8	EXIST::FUNCTION:CMP
 OSSL_CMP_MSG_update_recipNonce          5559	3_0_9	EXIST::FUNCTION:CMP
 ossl_ctx_legacy_digest_signatures_allowed ?	3_0_1	EXIST::FUNCTION:
